Marker/Marker/Parser/MarkdownParser.swift

```diff
@@ -22,54 +22,97 @@ internal struct MarkdownParser {
+    /// Parses specified string and returns a tuple containing string stripped of symbols and an array of Markdown elements.
     ///
     /// - Parameter string: String to be parsed.
     /// - Returns: Tuple containing string stripped of tag characters and an array of Markdown elements.
     /// - Throws: Parser error.
     static func parse(_ string: String) throws -> (strippedString: String, elements: [MarkdownElement]) {
         guard
-            let underscoreEmSymbol = underscoreEmSymbol,
-            let asteriskEmSymbol = asteriskEmSymbol,
             let underscoreStrongSymbol = underscoreStrongSymbol,
             let asteriskStrongSymbol = asteriskStrongSymbol,
             let tildeStrikethroughSymbol = tildeStrikethroughSymbol,
             let equalUnderlineSymbol = equalUnderlineSymbol else {
                 return (string, [])
         }
-        
-        func transformToMarkdownElement(_ element: Element) throws -> MarkdownElement {
-            switch element.symbol {
-            case underscoreEmSymbol, asteriskEmSymbol:
-                return .em(element.range)
-            case underscoreStrongSymbol, asteriskStrongSymbol:
-                return .strong(element.range)
-            case tildeStrikethroughSymbol:
-                return .strikethrough(element.range)
-            case equalUnderlineSymbol:
-                return .underline(element.range)
+
+        let underscoreEmRule = Rule(symbol: underscoreEmSymbol)
+        let asteriskEmRule = Rule(symbol: asteriskEmSymbol)
+        let underscoreStrongRule = Rule(symbol: underscoreStrongSymbol)
+        let asteriskStrongRule = Rule(symbol: asteriskStrongSymbol)
+        let tildeStrikethroughRule = Rule(symbol: tildeStrikethroughSymbol)
+        let equalUnderlineRule = Rule(symbol: equalUnderlineSymbol)
+        let linkTextRule = Rule(openingSymbol: linkTextOpeningSymbol, closingSymbol: linkTextClosingSymbol)
+        let linkURLRule = Rule(openingSymbol: linkURLOpeningSymbol, closingSymbol: linkURLClosingSymbol)
+
+        let tokens = try TokenParser.parse(string,
+                                           using: [underscoreEmRule,
+                                                   asteriskEmRule,
+                                                   underscoreStrongRule,
+                                                   asteriskStrongRule,
+                                                   tildeStrikethroughRule,
+                                                   equalUnderlineRule,
+                                                   linkTextRule,
+                                                   linkURLRule])
+
+        guard tokens.count > 0 else {
+            return (string, [])
+        }
+
+        var strippedString = ""
+        var elements: [MarkdownElement] = []
+
+        var i = 0
+        while i < tokens.count {
```

Comment:
is there a specific reason you used a while loop as opposed to `for (index, element) in <collection>.enumerated()`?

i.e. this would remove the need to increment `i` each time
```
for (index, token) in tokens.enumerated() {
    // Process each token here
}
```

Reply:
I am using a while loop here because, in the case of parsing links, I am iterating through two tokens at once; the enumerated for loop wouldn't let me do the same.